### ISSUE TITLE
GPU-enabled builds should not call GPU APIs without --gpu

### ISSUE THREAD

**Issue by olupton** [BlueBrain/CoreNeuron#599]
#595 closes #345, because GPU-enabled builds using Random123 can now be executed on CPU.

When a GPU-enabled build is executed **without** `--gpu` but on a machine that **does** have a GPU, various data transfers and synchronisation calls are made, which can be seen e.g. with `nvprof`.

We should fix these and make sure all GPU/OpenACC operations are conditional on `nt->compute_gpu`.

**Comment by olupton** [comment]
A more problematic aspect of this is that `__managed__` is not safe to use on nodes without GPUs, as demonstrated by
```
#include <iostream>
__managed__ int x = 42;
int main() {
  std::cout << "x = " << x << std::endl;
  return 0;
}
```
which will segfault on a non-GPU node and print 42 on a node with an NVIDIA GPU.
The real-world example of this is
```
(gdb) bt
#0  0x0000000000570a65 in coreneuron::nrnran123_set_globalindex(unsigned int) ()
#1  0x0000000000496a8d in coreneuron::set_globals (path=<optimized out>, cli_global_seed=0 '\000', cli_global_seed_value=-1) at ../external/coreneuron/coreneuron/io/global_vars.cpp:134
#2  0x0000000000487c2a in coreneuron::nrn_init_and_load_data (argc=<optimized out>, argv=<optimized out>, checkPoints=0x7fffffffd7c0, is_mapping_needed=0 '\000', run_setup_cleanup=1 '\001')
    at ../external/coreneuron/coreneuron/apps/main1.cpp:210
#3  0x000000000046e68c in run_solve_core (argc=<optimized out>, argv=<optimized out>) at ../external/coreneuron/coreneuron/apps/main1.cpp:496
#4  0x000000000040eac0 in solve_core (argc=3, argv=0x7fffffffda58) at /gpfs/bbp.cscs.ch/home/olupton/nrn/build_ollimod2c/share/coreneuron/enginemech.cpp:49
#5  0x000000000040ea99 in main (argc=10867584, argv=0x57f240 <cudart::createGlobalState()>) at /gpfs/bbp.cscs.ch/home/olupton/nrn/build_ollimod2c/share/coreneuron/coreneuron.cpp:14
```
so a follow-up to https://github.com/BlueBrain/CoreNeuron/pull/595 will be needed to enable non-GPU execution in GPU builds _on nodes without GPUs._

**Comment by olupton** [comment]
I think https://github.com/BlueBrain/CoreNeuron/pull/795 et al. improve the situation here, I just checked in an NMODL + OpenACC build and did not see any GPU activity in non-GPU tests.

**Comment by olupton** [comment]
I see the same in a MOD2C build with those branches, so I will mark this issue as closed by #795.
